events.go

```diff
@@ -293,3 +297,118 @@ func convertL3CacheInfo(i *intelrdt.L3CacheInfo) *l3CacheInfo {
 		NumClosids: i.NumClosids,
 	}
 }
+
+func resolveContainers(context *cli.Context) ([]libcontainer.Container, error) {
+	var containers []libcontainer.Container
+	errs := []error{}
+	for _, cID := range context.Args() {
+		container, err := getContainerWithID(cID, context)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		containers = append(containers, container)
+	}
+
+	if len(errs) != 0 {
+		return nil, combineErrors(errs)
+	}
+	return containers, nil
+}
+
+func runForAllContainers(containers []libcontainer.Container, routine func(ctr libcontainer.Container) error) chan error {
+	errs := make(chan error, len(containers))
+	for _, c := range containers {
+		go func(ctr libcontainer.Container) {
+			err := routine(ctr)
+			if err != nil {
+				errs <- err
+			}
+		}(c)
+	}
+	return errs
+}
+
+func runForAllContainersSync(containers []libcontainer.Container, routine func(ctr libcontainer.Container) error) error {
+	group := &sync.WaitGroup{}
+	group.Add(len(containers))
+
+	errorsChannel := runForAllContainers(containers, func(ctr libcontainer.Container) error {
+		defer group.Done()
+		return routine(ctr)
+	})
+	group.Wait()
+
+	errs := []error{}
+	for i := 0; i < len(errorsChannel); i++ {
+		errs = append(errs, <-errorsChannel)
+	}
+
+	return combineErrors(errs)
+}
+
+func combineErrors(errs []error) error {
+	if len(errs) == 0 {
+		return nil
+	}
+
+	errorMessages := []string{}
+	for _, err := range errs {
+		errorMessages = append(errorMessages, err.Error())
+	}
+
+	return errors.New(strings.Join(errorMessages, "\n"))
+}
+
+func forwardOOMNotifications(containers []libcontainer.Container, events chan *event) chan struct{} {
+	group := &sync.WaitGroup{}
+	group.Add(len(containers))
+
+	done := make(chan struct{})
+	go func() {
+		group.Wait()
+		close(done)
+	}()
+
+	signal := make(chan struct{})
+	for _, c := range containers {
+		go func(ctr libcontainer.Container) {
+			defer group.Done()
+			oomNotification, err := ctr.NotifyOOM()
+			if err != nil {
+				logrus.Error(err)
+				close(signal)
+				return
+			}
+
+			for {
+				select {
+				case <-signal:
+					return
+				case _, ok := <-oomNotification:
+					if !ok {
+						// the channel was closed because the container stopped and the cgroups no longer exist.
+						close(signal)
```

Comment:
Why cancel all oom monitors when one container got stopped?

Reply:
At CloudFoundry we do not have a use case for `runc events --interval` that could drive the implementation to address questions like:
* Should runc exit when a container exits?
* Should runc keep running forever, or it should exit once all monitored containers exit? 

That is why we took the easiest implementation approach to make `runc` exit when one of the monitored container is gone. Happy to revisit the code if you could suggest some reasonable behaviour.

Reply:
I would expect that `runc` would exit when all of the containers you asked to monitor have died. Since this implementation already uses `sync.WaitGroup` it's odd that it's not done in the more straight-forward way IMO.

Reply:
I sort of feel like if I was running `runc events --interval 5s container1 container2 ... container100`, I would not want to keep track of "disappearing containers" that are no longer emitting metrics. Another view would be that I said "give me the events for all these containers" but now runc can't fulfil that, so bail out instead of confusing me.

I could also totally buy that it should carry on "best-effort".

Reply:
> I said "give me the events for all these containers" but now runc can't fulfil that

I would simply think runc is still fulfilling that, you can't get more events from some containers because they are exited and no events are generated from them any more.

`runc events` don't exit until all monitored containers are gone makes more sense to me.

Reply:
Alright cool. Happy to defer to both of you and fix this up. Thanks for the feedback.

Reply:
We've changed the behaviour so that one container stopping only prevents metrics from that container being emitted to `runc events`. Also, `runc events` exits when all containers exit.